tests/functional/ext/private_import/private_import.py

```diff
@@ -0,0 +1,120 @@
+# and not emit a message as long as this file has a parent directory called `pylint`, even though
+# we are not importing from that directory. (We would be importing from `pylint/pylint`.)
+from private_import import _private  # pylint: disable=import-self
+from private_import.other_file import _private
+from . import _private
+from astroid import _private  # [import-private-name]
+from sys import _private  # [import-private-name]
+
+# Ignore typecheck
+from typing import TYPE_CHECKING, List, Optional
+
+if TYPE_CHECKING:
+    import _TreeType
+    from types import _TreeType
+    from _types import TreeType
+    from _types import _TreeType
+
+# No error since imports are used as type annotations
+from classes import _PrivateClassA
+from classes import _PrivateClassB
+from classes import _PrivateClassC
+
+a: _PrivateClassA
+
+def b_func(class_b: _PrivateClassB):
+    print(class_b)
+
+def c_func() -> _PrivateClassC:
+    return None
+
+# Used as typing in slices
+from classes import _SubScriptA
+from classes import _SubScriptB
+
+a: Optional[_SubScriptA]
+b: Optional[_SubScriptB[List]]
+
+import _TypeContainerA
+import _TypeContainerB
+import _TypeContainerC
+
+a2: _TypeContainerA.A
+
+def b_func2(class_b2: _TypeContainerB.B):
+    print(class_b2)
+
+def c2_func() -> _TypeContainerC.C:
+    return None
+
+# This is allowed since all the imports are used for typing
+from _TypeContainerExtra import TypeExtraA, TypeExtraB
+extraA: TypeExtraA
+extraB: TypeExtraB
+
+# This is not allowed because there is an import not used for typing
+from _TypeContainerExtra2 import TypeExtra2, NotTypeExtra # [import-private-name]
+extra2: TypeExtra2
+
+# Try many cases to ensure that type annotation usages of a private import
+# do not mask other illegal usages of the import
+import _private_module # [import-private-name]
+my_var: _private_module.Thing = _private_module.Thing()
+
+import _private_module2 # [import-private-name]
+my_var2: _private_module2.Thing2
+my_var2 = _private_module2.Thing2()
+
+import _private_module3 # [import-private-name]
+my_var3: _private_module3.Thing3
+my_var3 = _private_module3.Thing3
+my_var3_2: _private_module3.Thing3
+
+import _private_module4 # [import-private-name]
+my_var4: _private_module4.Thing4
+my_var4 = _private_module4.get_callers().get_thing4()
+
+from _private_module5 import PrivateClass # [import-private-name]
+my_var5: PrivateClass
+my_var5 = PrivateClass()
+
+from _private_module6 import PrivateClass2 # [import-private-name]
+my_var6: PrivateClass2 = PrivateClass2()
+
+from public_module import _PrivateClass3 # [import-private-name]
+my_var7: _PrivateClass3 = _PrivateClass3()
+
+# In this final case, the type annotation masks the access of a private name
+# However, this depends on annotating a variable and never using it.
+# This would be strange and in function bodies emits `unused-variable`
+import _private_module_unreachable
+my_var8: _private_module_unreachable.Thing8
```

Comment:
Can we warn here as well, or is that too difficult to do? I don't really like not warning "because another message will also indicate that this is wrong". What if that message is disabled? Then this becomes a false negative.

Reply:
I can't think of an straightforward way to close this completely without scanning the whole file. 

We could get closer by counting unassigned annotated variables as invalid uses, which would function if `unused-variable` is disabled.

But this would still fail to the case
```
import public_thing8_maker
import _private_module_unreachable
my_var8: _private_module_unreachable.Thing8
my_var8 = public_thing8_maker.build_thing8()

my_var9 = _private_module_unreachable.Thing8()
```

But I do think we're making progress in these false negatives being increasingly improbable to see in real code.